File: Python/tests/test_adstock_model.py
import pytest
import numpy as np
import pandas as pd
from adstock_model import AdstockModel

@pytest.fixture
def sample_data():
    data = pd.DataFrame({
        "Date": pd.date_range(start="2020-01-01", periods=10, freq="W"),
        "EQ_Volume": np.random.normal(100, 20, 10),
        "Media_1": np.random.normal(50, 10, 10),
        "Media_2": np.random.normal(30, 5, 10)
    })
    return data

@pytest.fixture
def sample_saturation_models():
    return {
        "linear": lambda x: x,
        "diminishing": lambda x: x / (1 + x),
        "log": lambda x: np.log1p(x),
        "exponential": lambda x: 1 - np.exp(-x)
    }

@pytest.fixture
def sample_distribution_params():
    return {
        "Weibull": {"param1": [0.1, 5, 0.1], "param2": [1, 70, 1]},
        "Lognormal": {"param1": [1, 20, 1], "param2": [10, 20, 1]},
        "Logistic": {"param1": [1, 16, 1], "param2": [1, 16, 1]},
        "NegExp": {"param1": [0.1, 1, 0.1], "param2": [None, None, None]},
        "ChiSquare": {"param1": [1, 70, 1], "param2": [None, None, None]}
    }

@pytest.fixture
def adstock_model(sample_data, sample_saturation_models, sample_distribution_params):
    dates = sample_data["Date"]
    return AdstockModel(sample_data, dates, sample_saturation_models, sample_distribution_params)

def test_adstock_initialization(adstock_model):
    assert isinstance(adstock_model, AdstockModel)
    assert adstock_model.data is not None
    assert isinstance(adstock_model.saturation_models, dict)
    assert isinstance(adstock_model.distribution_params, dict)

def test_adstock_run(adstock_model):
    adstock_model.run_adstock()
    assert adstock_model.results is not None
    assert not adstock_model.results.empty

def test_compute_correlations(adstock_model):
    adstock_model.run_adstock()
    correlations = adstock_model.compute_correlations()
    assert isinstance(correlations, pd.DataFrame)
    assert "Variable" in correlations.columns
    assert "Correl" in correlations.columns
    assert "Pval" in correlations.columns

def test_lagpad_function(adstock_model):
    data = np.array([1, 2, 3, 4, 5])
    lagged_data = adstock_model.lagpad(data, 2)
    assert np.isnan(lagged_data[0]) and np.isnan(lagged_data[1])
    assert lagged_data[2] == 1

def test_normalization(adstock_model):
    data = np.array([1, 2, 3, 4, 5])
    normalized_data = adstock_model.normalize(data)
    assert np.isclose(np.sum(normalized_data), 1)

def test_ztran(adstock_model):
    data = np.array([1, 2, 3, 4, 5])
    transformed_data = adstock_model.ztran(data)
    assert np.isclose(np.mean(transformed_data), 0)
    assert np.isclose(np.std(transformed_data), 1)

if __name__ == "__main__":
    pytest.main()

File: R/tests/test_adstock_model.R
library(testthat)
library(dplyr)
library(data.table)

# Load the AdstockModel source file if needed
# source("adstock_model.R")

# Sample data for testing
sample_data <- data.frame(
  Date = seq(as.Date("2020-01-01"), by = "week", length.out = 10),
  EQ_Volume = rnorm(10, mean = 100, sd = 20),
  Media_1 = rnorm(10, mean = 50, sd = 10),
  Media_2 = rnorm(10, mean = 30, sd = 5)
)

sample_saturation_models <- list(
  linear = function(x) x,
  diminishing = function(x) x / (1 + x),
  log = function(x) log(1 + x),
  exponential = function(x) 1 - exp(-x)
)

sample_distribution_params <- list(
  Weibull = list(param1 = c(0.1, 5, 0.1), param2 = c(1, 70, 1)),
  Lognormal = list(param1 = c(1, 20, 1), param2 = c(10, 20, 1)),
  Logistic = list(param1 = c(1, 16, 1), param2 = c(1, 16, 1)),
  NegExp = list(param1 = c(0.1, 1, 0.1), param2 = c(NA, NA, NA)),
  ChiSquare = list(param1 = c(1, 70, 1), param2 = c(NA, NA, NA))
)

# Initialize Adstock Model
adstock_model <- AdstockModel(
  data = sample_data,
  dates = sample_data$Date,
  saturation_models = sample_saturation_models,
  distribution_params = sample_distribution_params
)

# Test initialization
test_that("Adstock model initializes correctly", {
  expect_true(is.list(adstock_model))
  expect_true("data" %in% names(adstock_model))
  expect_true("saturation_models" %in% names(adstock_model))
  expect_true("distribution_params" %in% names(adstock_model))
})

# Test running adstock
test_that("Adstock model runs without errors", {
  result <- runAdstock(adstock_model)
  expect_true(!is.null(result$results))
  expect_true(nrow(result$results) > 0)
})

# Test correlation computation
test_that("Correlation computation works correctly", {
  result <- runAdstock(adstock_model)
  correlations <- computeCorrelations(result)
  expect_true("Variable" %in% colnames(correlations))
  expect_true("Correl" %in% colnames(correlations))
  expect_true("Pval" %in% colnames(correlations))
})

# Test normalization function
test_that("Normalization function works correctly", {
  normalized_data <- normalize(c(1, 2, 3, 4, 5))
  expect_equal(sum(normalized_data), 1, tolerance = 1e-6)
})

# Test z-transformation function
test_that("Z-transformation standardizes correctly", {
  data <- c(1, 2, 3, 4, 5)
  transformed_data <- ztran(data)
  expect_equal(mean(transformed_data), 0, tolerance = 1e-6)
  expect_equal(sd(transformed_data), 1, tolerance = 1e-6)
})

# Run tests
test_file("test_adstock_model.R")
